tsl/src/compression/api.c

```diff
@@ -892,3 +896,584 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+			MakeTupleTableSlot(uncompressed_rel_tupdesc, &TTSOpsHeapTuple);
+
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+
+		tuplesort_puttupleslot(segment_tuplesortstate, heap_tuple_slot);
+
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+
+		ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	}
+	table_endscan(heapScan);
+}
+
+static void
+fetch_matching_uncompressed_chunk_into_tuplestore(Tuplesortstate *segment_tuplesortstate,
+												  int nsegmentby_cols,
+												  Relation uncompressed_chunk_rel,
+												  CompressedSegmentInfo **current_segment)
+{
+	TableScanDesc heapScan;
+	HeapTuple uncompressed_tuple;
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+	ScanKeyData *scankey = palloc0(sizeof(*scankey) * nsegmentby_cols);
+
+	for (int seg_col = 0; seg_col < nsegmentby_cols; seg_col++)
+	{
+		Datum val = current_segment[seg_col]->segment_info->val;
+		// what is the attno of this segby col in the uncompressed chunk?
+		int16 attno =
+			current_segment[seg_col]->decompressed_chunk_offset + 1; // offset is attno - 1
+		ScanKeyEntryInitializeWithInfo(&scankey[seg_col],
+									   0 /*flags*/,
+									   attno,
+									   BTEqualStrategyNumber,
+									   InvalidOid,
+									   current_segment[seg_col]->segment_info->collation,
+									   &current_segment[seg_col]->segment_info->eq_fn,
+									   val);
+	}
+
+	heapScan =
+		table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), nsegmentby_cols, scankey);
+
+	for (uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+		 uncompressed_tuple != NULL;
+		 uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	{
+		TupleTableSlot *heap_tuple_slot =
+			MakeTupleTableSlot(uncompressed_rel_tupdesc, &TTSOpsHeapTuple);
+
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(segment_tuplesortstate, heap_tuple_slot);
+		/* simple_heap_delete since we don't expect concurrent updates, have exclusive lock on the
+		 * relation */
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+		ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	}
+	table_endscan(heapScan);
+	pfree(scankey);
+}
+/*
+ * This function will replace the procedure for recompress_chunk
+ * It will keep the compressed chunk around, only decompressing the
+ * batches that are affected by the addition of newer data.
+ * If more batches need to be created, it will create those and update
+ * the meta_sequence_number of the following ones.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed =
+		PG_ARGISNULL(1) ? false : PG_GETARG_BOOL(1); // probably don't need this
```

Comment:
Please resolve this open question in the final version of the PR.

Reply:
I think it couldn't hurt to check this again, so decided to leave it there.